File: src/main/java/de/saschahlusiak/freebloks/model/Spielleiter.java
package de.saschahlusiak.freebloks.model;

import java.io.Serializable;
import java.util.Arrays;

import de.saschahlusiak.freebloks.controller.GameMode;

import static de.saschahlusiak.freebloks.controller.GameMode.*;

/**
 * Enriches the current board state with meta information about the game:
 *
 * - game state
 * - history for undo
 * - current player
 * - whether players are controlled locally or by the computer
 * - current game mode
 */
public class Spielleiter extends Spiel implements Serializable {
	private static final long serialVersionUID = -7880809258246268794L;

	public static final int PLAYER_COMPUTER = -2;
	public static final int PLAYER_LOCAL = -1;

	public int m_current_player;
	public int spieler[] = new int[Spiel.PLAYER_MAX];
	private GameMode m_gamemode;
	private boolean finished = false;
	private boolean started = false;

	public Turnpool history;

	public Spielleiter(int size) {
		super(size);
		m_current_player=-1;
		m_gamemode = GAMEMODE_4_COLORS_4_PLAYERS;
		for (int i=0;i<PLAYER_MAX;i++)spieler[i]=PLAYER_COMPUTER;
		history=new Turnpool();
	}

	public void set_noplayer() {
		m_current_player=-1;
	}

	public int current_player() {
		return m_current_player;
	}

	public Player get_current_player() {
		if (m_current_player == -1)
			return null;
		else
			return getPlayer(m_current_player);
	}

	public GameMode getGameMode() {
		return m_gamemode;
	}

	public void setGameMode(GameMode gameMode)
	{
		this.m_gamemode = gameMode;
	}

	public void addHistory(Turn turn) {
		history.add(new Turn(turn));
	}

	int num_players() {
		int n;
		n=0;
		for (int i=0;i<PLAYER_MAX;i++)if (spieler[i]!=PLAYER_COMPUTER)n++;
		return n;
	}

	/**
	 * Gibt true zurueck, wenn der Spieler kein Computerspieler ist
	 **/
	public boolean is_local_player(int player) {
		/*
		 * Bei keinem aktuellem Spieler, ist der aktuelle natuerlich nicht
		 * lokal.
		 */
		if (player == -1)
			return false;
		return (spieler[player] != Spielleiter.PLAYER_COMPUTER);
	}

	public boolean is_local_player() {
		return is_local_player(m_current_player);
	}

	public boolean isFinished() {
		return finished;
	}

	public void setFinished(boolean finished) {
		this.finished = finished;
	}

	public boolean isStarted() {
		return started;
	}

	public void setStarted(boolean started) {
		this.started = started;
	}

	public PlayerData[] getResultData() {
		PlayerData[] data;
		int i;
		switch (m_gamemode) {
		case GAMEMODE_2_COLORS_2_PLAYERS:
		case GAMEMODE_DUO:
		case GAMEMODE_JUNIOR:
			data = new PlayerData[2];
			data[0] = new PlayerData(this, 0);
			data[1] = new PlayerData(this, 2);
			break;

		case GAMEMODE_4_COLORS_2_PLAYERS:
			data = new PlayerData[2];
			data[0] = new PlayerData(this, 0, 2);
			data[1] = new PlayerData(this, 1, 3);
			break;

		case GAMEMODE_4_COLORS_4_PLAYERS:
		default:
			data = new PlayerData[4];
			data[0] = new PlayerData(this, 0);
			data[1] = new PlayerData(this, 1);
			data[2] = new PlayerData(this, 2);
			data[3] = new PlayerData(this, 3);
			break;
		}

		Arrays.sort(data);
		int place;
		for (i = 0; i < data.length; i++) {
			place = i + 1;
			if (i > 0) {
				if (data[i].compareTo(data[i-1]) == 0)
					place = data[i-1].getPlace();
			}

			data[i].setPlace(place);
		}
		return data;
	}
}

File: src/main/java/de/saschahlusiak/freebloks/model/Spielleiter.kt
package de.saschahlusiak.freebloks.model

import java.io.Serializable
import java.util.*

/**
 * Enriches the current board state with meta information about the game:
 *
 * - game state
 * - history for undo
 * - current player
 * - whether players are controlled locally or by the computer
 * - current game mode
 */
class Spielleiter(size: Int) : Spiel(size), Serializable {
    // -1 is "no current player", used in between states
    var currentPlayer = -1

    @JvmField
    val spieler = IntArray(PLAYER_MAX) { PLAYER_COMPUTER }
    var gameMode = GameMode.GAMEMODE_4_COLORS_4_PLAYERS
    var isFinished = false
    var isStarted = false
    val history = Turnpool()

    /**
     * @param player the player number to check or currentPlayer
     * @return true if the given player is played on the local device
     */
    @JvmOverloads
    fun isLocalPlayer(player: Int = currentPlayer): Boolean {
        return if (player == -1) false else spieler[player] != PLAYER_COMPUTER
    }

    fun getPlayerScores(): Array<PlayerScore> {
        val data: Array<PlayerScore>

        when (gameMode) {
            GameMode.GAMEMODE_2_COLORS_2_PLAYERS,
            GameMode.GAMEMODE_DUO,
            GameMode.GAMEMODE_JUNIOR -> {
                data = arrayOf(
                    PlayerScore(this, 0),
                    PlayerScore(this, 2)
                )
            }
            GameMode.GAMEMODE_4_COLORS_2_PLAYERS -> {
                data = arrayOf(
                    PlayerScore(this, 0, 2),
                    PlayerScore(this, 1, 3)
                )
            }
            GameMode.GAMEMODE_4_COLORS_4_PLAYERS -> {
                data = arrayOf(
                    PlayerScore(this, 0),
                    PlayerScore(this, 1),
                    PlayerScore(this, 2),
                    PlayerScore(this, 3)
                )
            }
        }

        Arrays.sort(data)
        // first everybody gets the natural place number
        data.forEachIndexed { index, p -> p.place = index + 1 }

        // then give everybody with equal points the same place
        for (i in 1 until data.size) {
            if (data[i].compareTo(data[i - 1]) == 0)
                data[i].place = data[i - 1].place
        }

        return data
    }

    companion object {
        const val PLAYER_COMPUTER = -2
        const val PLAYER_LOCAL = -1
    }
}